### ISSUE TITLE
Tests failing

### ISSUE THREAD

**Issue by Ghost-script** [mtconnect/cppagent#16]
While running the tests i am getting errors. Running on Fedora 21, gcc 4.9.2
Steps followed:

1. cmake . (in cppagent/)

2. make (in cppagent/)

3. ./agent_test (in cppgent/test/)

[subho@localhost] test $ (master) ./agent_test 
AdapterTest::testAdapter : OK
AgentTest::testConstructor  104 ERROR [0] xml.parser: XML: I/O 
  104 ERROR [0] xml.parser: XML: warning : 
  104 ERROR [0] xml.parser: XML: failed to load external entity "../samples/badPath.xml"

  104 FATAL [0] agent: Error loading xml configuration: ../samples/badPath.xml
  104 FATAL [0] agent: Error detail: XML Error at /home/subho/work/github/cppagent/agent/xml_parser.cpp(74): mDoc = xmlReadFile(aPath.c_str(), NULL, XML_PARSE_NOBLANKS)
XML Error at /home/subho/work/github/cppagent/agent/xml_parser.cpp(74): mDoc = xmlReadFile(aPath.c_str(), NULL, XML_PARSE_NOBLANKS)
 : OK
AgentTest::testBadPath : OK
AgentTest::testProbe : OK
AgentTest::testBadXPath  116 ERROR [0] xml.parser: XML: Invalid expression

  116 ERROR [0] xml.parser: XML: Invalid expression

  116 ERROR [0] xml.parser: XML: Invalid expression

  116 ERROR [0] xml.parser: XML: Invalid predicate

  116 ERROR [0] xml.parser: XML: Invalid expression

 : OK
AgentTest::testXPath : OK
AgentTest::testBadCount : OK
AgentTest::testBadFreq : OK
AgentTest::testGoodPath : OK
AgentTest::testEmptyStream : OK
AgentTest::testBadDevices : OK
AgentTest::testAddAdapter : OK
AgentTest::testAddToBuffer : OK
AgentTest::testAdapter : OK
AgentTest::testCurrentAt : OK
AgentTest::testCurrentAt64 : OK
AgentTest::testCurrentAtOutOfRange : OK
AgentTest::testSampleAtNextSeq : OK
AgentTest::testAdapterCommands : OK
AgentTest::testFileDownload : OK
AgentTest::testFailedFileDownload  410 ERROR [0] agent: Cannot register file: ./BadFileName.xsd: Unable to find file ./BadFileName.xsd
 : OK
AgentTest::testDuplicateCheck : OK
AgentTest::testAutoAvailable : OK
AgentTest::testIgnoreTimestamps : OK
AgentTest::testAssetStorage : OK
AgentTest::testAssetError : OK
AgentTest::testAssetBuffer : OK
AgentTest::testPut : OK
AgentTest::testPutBlocking : OK
AgentTest::testPutBlockingFrom : OK
AgentTest::testAdapterAddAsset : OK
AgentTest::testMultiLineAsset : OK
AgentTest::testAssetProbe : OK
AgentTest::testAssetStorageWithoutType : OK
AgentTest::testStreamData : OK
AgentTest::testSequenceNumberRollover : OK
AgentTest::testSampleCount : OK
AgentTest::testStreamDataObserver : OK
AgentTest::testFailWithDuplicateDeviceUUID 3665 ERROR [0] xml.parser: XML: I/O 
 3665 ERROR [0] xml.parser: XML: warning : 
 3665 ERROR [0] xml.parser: XML: failed to load external entity "../samples/dup_uuid.xml"

 3665 FATAL [0] agent: Error loading xml configuration: ../samples/dup_uuid.xml
 3665 FATAL [0] agent: Error detail: XML Error at /home/subho/work/github/cppagent/agent/xml_parser.cpp(74): mDoc = xmlReadFile(aPath.c_str(), NULL, XML_PARSE_NOBLANKS)
XML Error at /home/subho/work/github/cppagent/agent/xml_parser.cpp(74): mDoc = xmlReadFile(aPath.c_str(), NULL, XML_PARSE_NOBLANKS)
 : OK
AgentTest::testMultipleDisconnect : OK
AgentTest::testRelativeTime : OK
AgentTest::testRelativeParsedTime : OK
AgentTest::testRelativeParsedTimeDetection : OK
AgentTest::testRelativeOffsetDetection : OK
AgentTest::testDynamicCalibration : OK
AgentTest::testInitialTimeSeriesValues : OK
AgentTest::testUUIDChange : OK
AgentTest::testFilterValues : OK
AgentTest::testReferences : OK
AgentTest::testDiscrete : OK
AgentTest::testUpcaseValues : OK
AgentTest::testConditionSequence : OK
AgentTest::testAssetRemoval : OK
AgentTest::testAssetRemovalByAdapter : OK
AgentTest::testAssetAdditionOfAssetChanged12 : OK
AgentTest::testAssetAdditionOfAssetRemoved13 : OK
AgentTest::testAssetPrependId : OK
AgentTest::testBadAsset 3773 ERROR [0] xml.parser: XML: file://111.xml:1: 
 3773 ERROR [0] xml.parser: XML: parser 
 3773 ERROR [0] xml.parser: XML: error : 
 3773 ERROR [0] xml.parser: XML: Opening and ending tag mismatch: CuttingItem line 1 and CuttingItmems

 3773 ERROR [0] xml.parser: XML: <PointAngle code="SIG" nominal="0">0</PointAngle></Measurements></CuttingItmems>

 3773 ERROR [0] xml.parser: XML:                                                                                ^

 3773 ERROR [0] xml.parser: XML: file://111.xml:1: 
 3773 ERROR [0] xml.parser: XML: parser 
 3773 ERROR [0] xml.parser: XML: error : 
 3773 ERROR [0] xml.parser: XML: Opening and ending tag mismatch: CuttingItem line 1 and CuttingToolLifeCycle

 3773 ERROR [0] xml.parser: XML: nominal="0">0</PointAngle></Measurements></CuttingItmems></CuttingToolLifeCycle>

 3773 ERROR [0] xml.parser: XML:                                                                                ^

 3773 ERROR [0] xml.parser: XML: file://111.xml:1: 
 3773 ERROR [0] xml.parser: XML: parser 
 3773 ERROR [0] xml.parser: XML: error : 
 3773 ERROR [0] xml.parser: XML: Opening and ending tag mismatch: CuttingItem line 1 and CuttingTool

 3774 ERROR [0] xml.parser: XML: </PointAngle></Measurements></CuttingItmems></CuttingToolLifeCycle></CuttingTool

 3774 ERROR [0] xml.parser: XML:                                                                                ^

 3774 ERROR [0] xml.parser: XML: file://111.xml:2: 
 3774 ERROR [0] xml.parser: XML: parser 
 3774 ERROR [0] xml.parser: XML: error : 
 3774 ERROR [0] xml.parser: XML: Premature end of data in tag CuttingItems line 1

 3774 ERROR [0] xml.parser: XML: 

 3774 ERROR [0] xml.parser: XML: ^

 3774 ERROR [0] xml.parser: XML: file://111.xml:2: 
 3774 ERROR [0] xml.parser: XML: parser 
 3774 ERROR [0] xml.parser: XML: error : 
 3774 ERROR [0] xml.parser: XML: Premature end of data in tag CuttingToolLifeCycle line 1

 3774 ERROR [0] xml.parser: XML: 

 3774 ERROR [0] xml.parser: XML: ^

 3774 ERROR [0] xml.parser: XML: file://111.xml:2: 
 3774 ERROR [0] xml.parser: XML: parser 
 3774 ERROR [0] xml.parser: XML: error : 
 3774 ERROR [0] xml.parser: XML: Premature end of data in tag CuttingTool line 1

 3774 ERROR [0] xml.parser: XML: 

 3774 ERROR [0] xml.parser: XML: ^

 3774 ERROR [0] xml.parser: Cannot parse asset XML, Unknown execption occurred
 3774 ERROR [0] agent: addAssete: Error parsing asset
 : OK
AgentTest::testAssetWithSimpleCuttingItems : OK
AgentTest::testRemoveLastAssetChanged : OK
AgentTest::testRemoveAllAssets : OK
AgentTest::testEmptyLastItemFromAdapter : OK
AgentTest::testAdapterDeviceCommand 3798 ERROR [0] xml.parser: XML: I/O 
 3798 ERROR [0] xml.parser: XML: warning : 
 3798 ERROR [0] xml.parser: XML: failed to load external entity "../samples/two_devices.xml"

 3798 FATAL [0] agent: Error loading xml configuration: ../samples/two_devices.xml
 3798 FATAL [0] agent: Error detail: XML Error at /home/subho/work/github/cppagent/agent/xml_parser.cpp(74): mDoc = xmlReadFile(aPath.c_str(), NULL, XML_PARSE_NOBLANKS)
XML Error at /home/subho/work/github/cppagent/agent/xml_parser.cpp(74): mDoc = xmlReadFile(aPath.c_str(), NULL, XML_PARSE_NOBLANKS)
Segmentation fault (core dumped)

**Comment by wsobel** [comment]
What directory are you running the test from? It must be run from the test source directory.

**Comment by Ghost-script** [comment]
I am running it from cppagent/test

**Comment by wsobel** [comment]
Did you try debugging the set fault and see where it is occurring?

**Comment by Ghost-script** [comment]
gdb run ends with this. 

Program received signal SIGSEGV, Segmentation fault.
0x0000003368bb81b8 in main_arena () from /lib64/libc.so.6

Also a backtrace report is here:

(gdb) bt
#0  0x0000003368bb81b8 in main_arena () from /lib64/libc.so.6
#1  0x00000000005f2bce in ?? ()
#2  0x00007fffffffd8a0 in ?? ()
#3  0x0000000000989150 in ?? ()
#4  0x00007fffffffd840 in ?? ()
#5  0x0000000000642afb in ?? ()
#6  0x00007fffffffda60 in ?? ()
#7  0x00000000009890c0 in ?? ()
#8  0x00007fffffffda60 in ?? ()
#9  0x00000033684233a2 in CppUnit::TestCaseMethodFunctor::operator() (
    this=<optimized out>) at TestCase.cpp:32
#10 0x0000003368418f73 in CppUnit::DefaultProtector::protect (this=0x97c3d0, 
    functor=..., context=...) at DefaultProtector.cpp:15
#11 0x000000336842038d in operator() (this=<optimized out>)
    at ProtectorChain.cpp:20
#12 CppUnit::ProtectorChain::protect (this=0x97c110, functor=..., context=...)
    at ProtectorChain.cpp:77
#13 0x0000003368429459 in CppUnit::TestResult::protect (
    this=this@entry=0x7fffffffdcb0, functor=..., test=test@entry=0x9890c0, 
    shortDescription="tearDown() failed") at TestResult.cpp:178
#14 0x0000003368423110 in CppUnit::TestCase::run (this=0x9890c0, 
    result=0x7fffffffdcb0) at TestCase.cpp:97
#15 0x0000003368423743 in CppUnit::TestComposite::doRunChildTests (
---Type <return> to continue, or q <return> to quit---
    this=0x97ca20, controller=0x7fffffffdcb0) at TestComposite.cpp:64
#16 0x000000336842365e in CppUnit::TestComposite::run (this=0x97ca20, 
    result=0x7fffffffdcb0) at TestComposite.cpp:23
#17 0x0000003368423743 in CppUnit::TestComposite::doRunChildTests (
    this=0x97c960, controller=0x7fffffffdcb0) at TestComposite.cpp:64
#18 0x000000336842365e in CppUnit::TestComposite::run (this=0x97c960, 
    result=0x7fffffffdcb0) at TestComposite.cpp:23
#19 0x0000003368429022 in CppUnit::TestResult::runTest (this=0x7fffffffdcb0, 
    test=0x97c900) at TestResult.cpp:145
#20 0x000000336842ba4e in CppUnit::TestRunner::run (this=<optimized out>, 
    controller=..., testPath=...) at TestRunner.cpp:96
#21 0x00000000005264a0 in ?? ()
#22 0x00007fffffffdf38 in ?? ()
#23 0x00000001ffffdcbf in ?? ()
#24 0x0000000000000028 in ?? ()
#25 0x0000003000000003 in ?? ()
#26 0x0000000000000000 in ?? ()

**Comment by wsobel** [comment]
Looks like a stack corruption. What platform are you testing on?

**Comment by shakthimaan** [comment]
The platform is mentioned as Fedora 21 and GNU Compiler Collection 4.9.2. I am able to reproduce the segmentation fault on Fedora 20 with GNU Compiler Collection 4.8.2.

**Comment by wsobel** [comment]
Probably compiler related. Not sure if it is a real problem or an artifact of the version. We should do some additional testing. I'll check on os x with some memory profiling.

**Comment by Ghost-script** [comment]
@wsobel, I git reset the commits one by one and found the last commit hash that compiled successfully and tests ran correctly on my machine is  0a7ba0c4291d374cc7bc353df592c7a51884a186.

Also I noticed AgentTest::testAdapterDeviceCommand test was not there in the above mentioned hash. Another thing I noticed is the error says it couldn't load the file two_devices.xml from samples folder, if you  look into the samples folder there is no such two_devices.xml file. Can that be the cause of the error?

AgentTest::testAdapterDeviceCommand 3798 ERROR [0] xml.parser: XML: I/O 
3798 ERROR [0] xml.parser: XML: warning : 
3798 ERROR [0] xml.parser: XML: failed to load external entity "../samples/two_devices.xml"

3798 FATAL [0] agent: Error loading xml configuration: ../samples/two_devices.xml
3798 FATAL [0] agent: Error detail: XML Error at /home/subho/work/github/cppagent/agent/xml_parser.cpp(74): mDoc = xmlReadFile(aPath.c_str(), NULL, XML_PARSE_NOBLANKS)
XML Error at /home/subho/work/github/cppagent/agent/xml_parser.cpp(74): mDoc = xmlReadFile(aPath.c_str(), NULL, XML_PARSE_NOBLANKS)
Segmentation fault (core dumped)
